iree/compiler/Dialect/Flow/Transforms/FoldCompatibleDispatchRegions.cpp

```diff
@@ -164,13 +164,18 @@ bool doesValueDependOnOperation(Value value, Operation *op) {
              value.getDefiningOp()->isBeforeInBlock(op)) {
     // Can't depend on |op| as it is defined prior to it.
     return false;
+  } else if (value.getDefiningOp()->getBlock() == op->getBlock() &&
+             !value.getDefiningOp()->isBeforeInBlock(op)) {
+    // Somehow depends because |op| is defined before one of |value| operands.
```

Comment:
this should not be possible (that's the whole point of SSA) - this indicates there's a bug elsewhere

Reply:
ah this is the termination case for the recursion - that's tricky

"somehow" is scary here as it indicates you don't know how it happened :)
Callout that this is the base case of the recursion.

You can also remove the todo - this entire file should be rewritten and any reordering is something we'd want to do in a separate pass :)

Reply:
I see, this does happen in one of added test case.

```
  func @notDominate() -> tensor<4xf32> {
    %c4 = constant 4 : index
    %0 = flow.variable.load @var1 : tensor<4xf32>
    %1 = flow.dispatch.region[%c4 : index](%arg0 = %0 : tensor<4xf32>) -> tensor<4xf32> {
      %4 = mhlo.add %arg0, %arg0 : tensor<4xf32>
      flow.return %4 : tensor<4xf32>
    }
    %2 = flow.variable.load @var2 : tensor<4xf32>
    %3 = flow.dispatch.region[%c4 : index](%arg0 = %0 : tensor<4xf32>, %arg1 = %2 : tensor<4xf32>) -> tensor<4xf32> {
      %4 = mhlo.subtract %arg1, %arg0 : tensor<4xf32>
      flow.return %4 : tensor<4xf32>
    }
    return %3 : tensor<4xf32>
  }
```

This is a simpler one, there is a more complicated case in ResNet50. I think we probably still need to mix reordering ops and merging blocks. Because the load op in the above test case could be another block. Things will get complicated. And yes, I think we should rewrite this at some point.

Reply:
fyi, before add termination case, the result will be something like:

```
  func @notDominate() -> tensor<4xf32> {
    %c4 = constant 4 : index
    %0 = flow.variable.load @var1 : tensor<4xf32>
    %1 = flow.dispatch.region[%c4 : index](%arg0 = %0 : tensor<4xf32>, %arg1 = %2 : tensor<4xf32>) -> tensor<4xf32> {
      %4 = mhlo.add %arg0, %arg0 : tensor<4xf32>
      %5 = mhlo.subtract %4, %arg1 : tensor<4xf32>
      flow.return %4 : tensor<4xf32>
    }
    %2 = flow.variable.load @var2 : tensor<4xf32>
    return %3 : tensor<4xf32>
  }
```

The dispatch op takes `%2` as input.

This can't be resolved if we fold it in a reverse way, ie merging `lhs` into `rhs` to make `flow.varaible.load` on the top. Because it's possible to have a op in the middle that takes the result of first `flow.dispatch.region` as input. I think this is something we want to tune in the future, like duplicate the computation or just stop merging in this kind of case. (For duplicating the computation, I mean to remain the first `flow.dispatch.region` and also copy the block to the second `flow.dispatch.region`.)

Reply:
👍 awesome!